`what4-abc` no longer builds after recent `what4-domains` deprecations

Attempting to build `what4-abc` on the latest commit (fa69134465e7afcf99d36667f807485a340920be) fails with:

```
$ cabal build what4-abc
Configuration is affected by the following files:
- cabal.project
Build profile: -w ghc-9.4.8 -O1
In order, the following will be built (use -v for more details):
 - what4-abc-0.1 (lib) (first run)
Preprocessing library for what4-abc-0.1...
Building library for what4-abc-0.1...
[1 of 1] Compiling What4.Solver.ABC ( src/What4/Solver/ABC.hs, /home/rscott/Documents/Hacking/Haskell/what4/dist-newstyle/build/x86_64-linux/ghc-9.4.8/what4-abc-0.1/build/What4/Solver/ABC.o, /home/rscott/Documents/Hacking/Haskell/what4/dist-newstyle/build/x86_64-linux/ghc-9.4.8/what4-abc-0.1/build/What4/Solver/ABC.dyn_o )

src/What4/Solver/ABC.hs:97:1: error: [-Wdeprecations, -Werror=deprecations]
    Module ‘What4.Utils.BVDomain’ is deprecated:
      "Use What4.Domains.BV instead"
   |
97 | import qualified What4.Utils.BVDomain as BVD
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/What4/Solver/ABC.hs:98:1: error: [-Wdeprecations, -Werror=deprecations]
    Module ‘What4.Utils.BVDomain.Arith’ is deprecated:
      "Use What4.Domains.BV.Arith instead"
   |
98 | import qualified What4.Utils.BVDomain.Arith as A
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/What4/Solver/ABC.hs:99:1: error: [-Wdeprecations, -Werror=deprecations]
    Module ‘What4.Utils.BVDomain.Bitwise’ is deprecated:
      "Use What4.Domains.BV.Bitwise instead"
   |
99 | import qualified What4.Utils.BVDomain.Bitwise as B
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: [Cabal-7125]
Failed to build what4-abc-0.1.
```

While not fatal, there are also a variety of domains-related warnings that arise when building the `what4` library:

```
[42 of 90] Compiling What4.Utils.AbstractDomains ( src/What4/Utils/AbstractDomains.hs, /home/rscott/Documents/Hacking/Haskell/what4/dist-newstyle/build/x86_64-linux/ghc-9.4.8/what4-1.7.3.0.99/build/What4/Utils/AbstractDomains.o, /home/rscott/Documents/Hacking/Haskell/what4/dist-newstyle/build/x86_64-linux/ghc-9.4.8/what4-1.7.3.0.99/build/What4/Utils/AbstractDomains.dyn_o )

src/What4/Utils/AbstractDomains.hs:695:24: warning: [-Wdeprecations]
    In the use of ‘any’ (imported from What4.Domains.BV):
    Deprecated: "Use 'top' instead"
    |
695 |     BaseBVRepr w    -> BVD.any w
    |                        ^^^^^^^

src/What4/Utils/AbstractDomains.hs:771:27: warning: [-Wdeprecations]
    In the use of ‘union’ (imported from What4.Domains.BV):
    Deprecated: "Use 'join' instead"
    |
771 |   avJoin (BaseBVRepr _) = BVD.union
    |                           ^^^^^^^^^
```
```
[58 of 90] Compiling What4.Expr.App   ( src/What4/Expr/App.hs, /home/rscott/Documents/Hacking/Haskell/what4/dist-newstyle/build/x86_64-linux/ghc-9.4.8/what4-1.7.3.0.99/build/What4/Expr/App.o, /home/rscott/Documents/Hacking/Haskell/what4/dist-newstyle/build/x86_64-linux/ghc-9.4.8/what4-1.7.3.0.99/build/What4/Expr/App.dyn_o )

src/What4/Expr/App.hs:2099:23: warning: [-Wdeprecations]
    In the use of ‘any’ (imported from What4.Domains.BV):
    Deprecated: "Use 'top' instead"
     |
2099 |       BaseBVRepr w -> BVD.any w
     |                       ^^^^^^^

src/What4/Expr/App.hs:2103:24: warning: [-Wdeprecations]
    In the use of ‘any’ (imported from What4.Domains.BV):
    Deprecated: "Use 'top' instead"
     |
2103 |     FloatToBV w _ _ -> BVD.any w
     |                        ^^^^^^^

src/What4/Expr/App.hs:2104:25: warning: [-Wdeprecations]
    In the use of ‘any’ (imported from What4.Domains.BV):
    Deprecated: "Use 'top' instead"
     |
2104 |     FloatToSBV w _ _ -> BVD.any w
     |                         ^^^^^^^
```

--- a/what4-abc/src/What4/Solver/ABC.hs
+++ b/what4-abc/src/What4/Solver/ABC.hs
@@ -79,6 +79,9 @@ import           System.Process
 import           What4.BaseTypes
 import           What4.Concrete
 import           What4.Config
+import qualified What4.Domains.BV as BVD
+import qualified What4.Domains.BV.Arith as A
+import qualified What4.Domains.BV.Bitwise as B
 import           What4.Interface
                    ( getConfiguration, IsExprBuilder, logSolverEvent
                    , SolverEvent(..), SolverStartSATQuery(..), SolverEndSATQuery(..), andAllOf )
@@ -94,9 +97,6 @@ import           What4.ProgramLoc
 import           What4.Solver.Adapter
 import           What4.SatResult
 import           What4.Utils.AbstractDomains
-import qualified What4.Utils.BVDomain as BVD
-import qualified What4.Utils.BVDomain.Arith as A
-import qualified What4.Utils.BVDomain.Bitwise as B
 import           What4.Utils.Complex
 import qualified What4.Utils.Environment as Env
 import           What4.Utils.MonadST

--- a/what4-abc/what4-abc.cabal
+++ b/what4-abc/what4-abc.cabal
@@ -32,9 +32,10 @@ library
     transformers,
     unordered-containers,
     utf8-string,
-    -- what4 lives in the same repo and this package isn't released
-    -- separately; no version bound is needed.
-    what4
+    -- what4 and what4-domains live in the same repo and this package isn't
+    -- released separately; no version bound is needed.
+    what4,
+    what4-domains
 
   hs-source-dirs: src
 

--- a/what4/src/What4/Expr/App.hs
+++ b/what4/src/What4/Expr/App.hs
@@ -2096,12 +2096,12 @@ abstractEval f a0 = do
     FloatRound{} -> ()
     FloatFromBinary{} -> ()
     FloatToBinary fpp _ -> case floatPrecisionToBVType fpp of
-      BaseBVRepr w -> BVD.any w
+      BaseBVRepr w -> BVD.top w
     BVToFloat{} -> ()
     SBVToFloat{} -> ()
     RealToFloat{} -> ()
-    FloatToBV w _ _ -> BVD.any w
-    FloatToSBV w _ _ -> BVD.any w
+    FloatToBV w _ _ -> BVD.top w
+    FloatToSBV w _ _ -> BVD.top w
     FloatToReal{} -> ravUnbounded
     FloatSpecialFunction{} -> ()
 

--- a/what4/src/What4/Utils/AbstractDomains.hs
+++ b/what4/src/What4/Utils/AbstractDomains.hs
@@ -692,7 +692,7 @@ avTop tp =
     BaseRealRepr    -> ravUnbounded
     BaseComplexRepr -> ravUnbounded :+ ravUnbounded
     BaseStringRepr _ -> stringAbsTop
-    BaseBVRepr w    -> BVD.any w
+    BaseBVRepr w    -> BVD.top w
     BaseFloatRepr{} -> ()
     BaseArrayRepr _a b -> avTop b
     BaseStructRepr flds -> fmapFC (\etp -> AbstractValueWrapper (avTop etp)) flds
@@ -768,7 +768,7 @@ instance Abstractable BaseRealType where
 
 -- Bitvectors always have a lower and upper bound (represented as unsigned numbers)
 instance (1 <= w) => Abstractable (BaseBVType w) where
-  avJoin (BaseBVRepr _) = BVD.union
+  avJoin (BaseBVRepr _) = BVD.join
   avOverlap _ = BVD.domainsOverlap
   avCheckEq _ = BVD.eq
 

